lib/install/management/appliance.go

```diff
@@ -139,6 +140,53 @@ func (d *Dispatcher) checkExistence(conf *config.VirtualContainerHostConfigSpec,
 	return err
 }
 
+func (d *Dispatcher) powerOffVM(vm *vm.VirtualMachine) error {
+	var err error
+	power, err := vm.PowerState(d.op)
+	if err == nil && power == types.VirtualMachinePowerStatePoweredOff {
+		d.op.Debugf("VM is already powered off: %s", vm.Reference())
+		return nil
+	}
+
+	if err != nil {
+		d.op.Warnf("Failed to get vm power status %q: %s", vm.Reference(), err)
+	}
+
+	// try guest shutdown first
+	tools, err := vm.IsToolsRunning(d.op)
+	if tools {
+		d.op.Debugf("Performing guest shutdown for %s", vm.Reference())
+		err = vm.ShutdownGuest(d.op)
+		if err == nil {
+			// just enough time for the endpoint to shutdown cleanly even having timed out internally
+			timeout, cancel := trace.WithTimeout(&d.op, shared.GuestShutdownTimeout+(5*time.Second), "Shut down endpointVM")
+			err = vm.WaitForPowerState(timeout, types.VirtualMachinePowerStatePoweredOff)
+			cancel()
+		}
+	}
+
+	if err != nil || !tools {
```

Comment:
The way this is currently written, I find myself wondering things like "In what cases will we hit this block?" (Where was `err` set? In what cases will/won't it be `nil`? Have  we already logged the `err`?)

I don't have a specific suggestion for how to improve this method, but I don't think the current structure is the clearest way to communicate the intended flow.